client/tracing/src/logging/mod.rs

```diff
@@ -0,0 +1,531 @@
+		.add_directive(
+			parse_default_directive("cranelift_codegen=off").expect("provided directive is valid"),
+		)
+		// Set warn logging by default for some modules.
+		.add_directive(
+			parse_default_directive("cranelift_wasm=warn").expect("provided directive is valid"),
+		)
+		.add_directive(parse_default_directive("hyper=warn").expect("provided directive is valid"));
+
+	if let Ok(lvl) = std::env::var("RUST_LOG") {
+		if lvl != "" {
+			env_filter = parse_user_directives(env_filter, &lvl)?;
+		}
+	}
+
+	if pattern != "" {
+		// We're not sure if log or tracing is available at this moment, so silently ignore the
+		// parse error.
+		env_filter = parse_user_directives(env_filter, pattern)?;
+	}
+
+	let max_level_hint = Layer::<FmtSubscriber>::max_level_hint(&env_filter);
+
+	let max_level = max_level.unwrap_or_else(|| match max_level_hint {
+		Some(tracing_subscriber::filter::LevelFilter::INFO) | None => log::LevelFilter::Info,
+		Some(tracing_subscriber::filter::LevelFilter::TRACE) => log::LevelFilter::Trace,
+		Some(tracing_subscriber::filter::LevelFilter::WARN) => log::LevelFilter::Warn,
+		Some(tracing_subscriber::filter::LevelFilter::ERROR) => log::LevelFilter::Error,
+		Some(tracing_subscriber::filter::LevelFilter::DEBUG) => log::LevelFilter::Debug,
+		Some(tracing_subscriber::filter::LevelFilter::OFF) => log::LevelFilter::Off,
+	});
+
+	tracing_log::LogTracer::builder()
+		.with_max_level(max_level)
+		.init()?;
+
+	// If we're only logging `INFO` entries then we'll use a simplified logging format.
+	let simple = match max_level_hint {
+		Some(level) if level <= tracing_subscriber::filter::LevelFilter::INFO => true,
+		_ => false,
+	};
+
+	let enable_color = force_colors.unwrap_or_else(|| atty::is(atty::Stream::Stderr));
+	let timer = ChronoLocal::with_format(if simple {
+		"%Y-%m-%d %H:%M:%S".to_string()
+	} else {
+		"%Y-%m-%d %H:%M:%S%.3f".to_string()
+	});
+
+	let (telemetry_layer, telemetry_worker) =
+		sc_telemetry::TelemetryLayer::new(telemetry_buffer_size, telemetry_external_transport)?;
+	let event_format = EventFormat {
+		timer,
+		display_target: !simple,
+		display_level: !simple,
+		display_thread_name: !simple,
+		enable_color,
+	};
+	let builder = FmtSubscriber::builder().with_env_filter(env_filter);
+
+	#[cfg(not(target_os = "unknown"))]
+	let builder = builder.with_writer(std::io::stderr as _);
+
+	#[cfg(target_os = "unknown")]
+	let builder = builder.with_writer(std::io::sink);
+
+	#[cfg(not(target_os = "unknown"))]
+	let builder = builder.event_format(event_format);
+
+	#[cfg(not(target_os = "unknown"))]
+	let builder = builder_hook(builder);
+
+	let subscriber = builder.finish().with(PrefixLayer).with(telemetry_layer);
+
+	#[cfg(target_os = "unknown")]
+	let subscriber = subscriber.with(ConsoleLogLayer::new(event_format));
+
+	Ok((subscriber, telemetry_worker))
+}
+
+/// A builder that is used to initialize the global logger.
+pub struct GlobalLoggerBuilder {
```

Comment:
Unless we have a "local" builder I think this is better named `LoggerBuilder`.

Reply:
I believe it is important to insist that this will be set globally for the whole process. It also sounds closer to tracing's API: https://docs.rs/tracing/0.1.21/tracing/subscriber/fn.set_global_default.html

Reply:
I disagree mildly, but definitely not a big deal. :)

Reply:
It doesn't have to be my way we can also ask for a third opinion :grin:

Reply:
I also was not very happy with the name in my review :P However, if there is now a poll, I'm voting for `LoggerBuilder`

Reply:
`LoggerBuilder` it is then :grin: